xdsl/dialects/csl_/csl_stencil.py

```diff
@@ -0,0 +1,271 @@
+from itertools import pairwise
+from typing import Annotated, cast
+
+from xdsl.dialects import builtin, memref, stencil
+from xdsl.dialects.builtin import IntegerAttr, IntegerType, TensorType
+from xdsl.dialects.experimental import dmp
+from xdsl.ir import Attribute, Dialect, Operation, ParametrizedAttribute, SSAValue
+from xdsl.irdl import (
+    ConstraintVar,
+    IRDLOperation,
+    Operand,
+    ParameterDef,
+    attr_def,
+    irdl_attr_definition,
+    irdl_op_definition,
+    operand_def,
+    opt_attr_def,
+    opt_prop_def,
+    prop_def,
+    result_def,
+)
+from xdsl.parser import AttrParser, Parser
+from xdsl.printer import Printer
+from xdsl.traits import HasAncestor, Pure
+from xdsl.utils.exceptions import VerifyException
+from xdsl.utils.hints import isa
+
+
+@irdl_attr_definition
+class ExchangeDeclarationAttr(ParametrizedAttribute):
+    """
+    A simplified version of dmp.exchange, from which it should be lowered
+    """
+
+    name = "csl_stencil.exchange"
+
+    neighbor_param: ParameterDef[builtin.DenseArrayBase]
+
+    def __init__(
+        self,
+        neighbor: Sequence[int] | builtin.DenseArrayBase,
+    ):
+        data_type = builtin.i64
+        super().__init__(
+            [
+                (
+                    neighbor
+                    if isinstance(neighbor, builtin.DenseArrayBase)
+                    else builtin.DenseArrayBase.from_list(data_type, neighbor)
+                ),
+            ]
+        )
+
+    @classmethod
+    def from_dmp_exch_decl_attr(cls, src: dmp.ExchangeDeclarationAttr):
+        return cls(src.neighbor)
+
+    @property
+    def neighbor(self) -> tuple[int, ...]:
+        data = self.neighbor_param.as_tuple()
+        assert isa(data, tuple[int, ...])
+        return data
+
+    def print_parameters(self, printer: Printer) -> None:
+        printer.print_string(f"<to {list(self.neighbor)}>")
+
+    @classmethod
+    def parse_parameters(cls, parser: AttrParser) -> list[Attribute]:
+        parser.parse_characters("<")
+        parser.parse_characters("to")
+        to = parser.parse_comma_separated_list(
+            parser.Delimiter.SQUARE, parser.parse_integer
+        )
+        parser.parse_characters(">")
+
+        return [builtin.DenseArrayBase.from_list(builtin.i64, to)]
+
+
+@irdl_op_definition
+class PrefetchOp(IRDLOperation):
+    """
+    An op to indicate buffer prefetches.
+    """
+
+    name = "csl_stencil.prefetch"
+
+    input_stencil: Operand = operand_def(
+        stencil.TempType[Attribute] | memref.MemRefType[Attribute]
+    )
+
+    swaps: builtin.ArrayAttr[ExchangeDeclarationAttr] | None = opt_prop_def(
+        builtin.ArrayAttr[ExchangeDeclarationAttr]
+    )
+
+    size = prop_def(IntegerAttr[IntegerType])
+
+    topo: dmp.RankTopoAttr | None = opt_prop_def(dmp.RankTopoAttr)
+
+    result = result_def(memref.MemRefType)
+
+    def __init__(
+        self,
+        input_stencil: SSAValue | Operation,
+        size: IntegerAttr[IntegerType],
```

Comment:
Isn't there an implicit constraint here, that the resulting memref is always a `memref<${len(self.swaps}xtensor<${self.size}x??>>`?

I feel we don't need to pass both size and result type, as this is redundant information?

Some documentation on each attribute would definitely help people to understand this op much better.

Reply:
Excellent, removing it immediately! Thanks